### PR TITLE
Scanner rework.

### PR DESCRIPTION
**Pull request by zherczeg**
The scanner was an on-demand tool of the parser, which was triggered by certain statements. After the rework, the scanner runs only once, and collects all information. This information is passed to the parser.

### CODE REVIEW

**Comment by akosthekiss** [comment]
Please, share memory and run time information (before/after) as soon as they are available.

**Comment by zherczeg** [comment]
Results: The geometric part is totally wrong. Only the text growth by 1K.

This patch is a feature patch, not an optimization patch. This infrastructure will be needed for implementing let/const and destructuring assignments (and arrow functions with initializers, which will be in the final part of this patch). And expect more follow-up patches which increase memory consumption further.

Benchmark | Heap (bytes) | Stack (bytes) | Perf (sec) |
----: | ----: | ----: | ----: | 
3d-cube.js | 16376 -> 16376 : 0.000% | 15448 -> 15448 : 0.000% | 0.766 -> 0.765 : +0.196% | 
3d-raytrace.js | 121872 -> 121872 : 0.000% | 2740 -> 2740 : 0.000% | 0.920 -> 0.907 : **+1.380%** | 
access-binary-trees.js | 32752 -> 32752 : 0.000% | 3272 -> 3272 : 0.000% | 0.460 -> 0.451 : **+2.053%** | 
access-fannkuch.js | 4960 -> 4960 : 0.000% | 1624 -> 1648 : -1.478% | 1.786 -> 1.852 : -3.699% | 
access-nbody.js | 6264 -> 6264 : 0.000% | 1600 -> 1632 : -2.000% | 0.996 -> 0.982 : **+1.400%** | 
bitops-3bit-bits-in-byte.js | 712 -> 720 : -1.124% | 1568 -> 1592 : -1.531% | 0.466 -> 0.467 : -0.114% | 
bitops-bits-in-byte.js | 664 -> 664 : 0.000% | 1568 -> 1592 : -1.531% | 0.628 -> 0.628 : +0.063% | 
bitops-bitwise-and.js | 496 -> 496 : 0.000% | 1344 -> 1368 : -1.786% | 0.773 -> 0.773 : -0.084% | 
bitops-nsieve-bits.js | 98304 -> 98304 : 0.000% | 1568 -> 1592 : -1.531% | 1.079 -> 1.073 : +0.490% | 
controlflow-recursive.js | 888 -> 888 : 0.000% | 67720 -> 67720 : 0.000% | 0.317 -> 0.324 : -2.425% | 
crypto-aes.js | 33200 -> 33200 : 0.000% | 2360 -> 2360 : 0.000% | 0.774 -> 0.770 : +0.451% | 
crypto-md5.js | 106488 -> 106488 : 0.000% | 1856 -> 1856 : 0.000% | 0.515 -> 0.509 : **+1.181%** | 
crypto-sha1.js | 65528 -> 65528 : 0.000% | 1680 -> 1712 : -1.905% | 0.520 -> 0.513 : **+1.218%** | 
date-format-tofte.js | 16368 -> 16368 : 0.000% | 2048 -> 2048 : 0.000% | 0.698 -> 0.711 : -1.753% | 
date-format-xparb.js | 16368 -> 16368 : 0.000% | 2528 -> 2560 : -1.266% | 0.485 -> 0.484 : +0.404% | 
math-cordic.js | 1728 -> 1728 : 0.000% | 1568 -> 1592 : -1.531% | 1.075 -> 1.060 : **+1.380%** | 
math-partial-sums.js | 1240 -> 1240 : 0.000% | 1528 -> 1560 : -2.094% | 0.650 -> 0.640 : **+1.451%** | 
math-spectral-norm.js | 6096 -> 6096 : 0.000% | 1568 -> 1592 : -1.531% | 0.473 -> 0.460 : **+2.879%** | 
string-base64.js | 90800 -> 90800 : 0.000% | 1600 -> 1632 : -2.000% | 1.234 -> 1.242 : -0.636% | 
string-fasta.js | 8256 -> 8256 : 0.000% | 1552 -> 1576 : -1.546% | 1.191 -> 1.180 : +0.897% | 
Geometric mean: | -21.210% | +1.022% | -19.911% | 

Binary (bytes) | master(26d48c3e28) | patch(2315cb59c7) | Diff |
----: | ----: | ----: | ----: |
size | 128568 | 128568 | 0 bytes |
.text | 111516 | 112676 | +1160 bytes |

**Comment by zherczeg** [comment]
Import export support is added.

File: jerry-core/parser/js/js-scanner.c [hunk 1]
```diff
@@ -0,0 +1,1702 @@
...
+            continue;
+          }
+          break;
+        }
+        case SCAN_MODE_FUNCTION_ARGUMENTS:
+        {
+#if ENABLED (JERRY_ES2015_CLASS)
+          JERRY_ASSERT (stack_top == SCAN_STACK_BLOCK_STATEMENT
+                        || stack_top == SCAN_STACK_BLOCK_EXPRESSION
+                        || stack_top == SCAN_STACK_CLASS
```
Comment (rerobika) [inline jerry-core/parser/js/js-scanner.c]: How about only a single assert with a guarded `stack_top_p == SCAN_STACK_CLASS` ?

Comment (zherczeg) [reply jerry-core/parser/js/js-scanner.c]: Because ASSERT is a macro, you cannot put guard into it. Some compilers support it but it is non-standard.

**Pull request merged**